src/peft/tuners/reft/layer.py

```diff
@@ -0,0 +1,195 @@
+        pass
+
+    def update_layer(
+        self,
+        adapter_name: str,
+        r: int,
+        alpha: float,
+        loc: Optional[str],
+        dropout: float,
+        **kwargs,
+    ) -> None:
+        """Internal function to create loha adapter
+
+        Args:
+            adapter_name (`str`): Name for the adapter to add.
+            r (`int`): Rank for the added adapter.
+            alpha (`float`): Alpha for the added adapter.
+            dropout (`float`): The dropout probability for disabling adapter during training.
+        """
+        if r <= 0:
+            raise ValueError(f"`r` should be a positive integer value but the value passed is {r}")
+
+        self.r[adapter_name] = r
+        first_n, last_n = parse_positions(loc) if loc else (0, 0)
+        self.first_n[adapter_name] = first_n
+        self.last_n[adapter_name] = last_n
+        self.alpha[adapter_name] = alpha
+        self.scaling[adapter_name] = alpha / r
+        self.dropout[adapter_name] = torch.nn.Dropout(dropout)
+
+        # Create weights with provided shape
+        self.create_adapter_parameters(adapter_name, r)
+
+        # Initialize weights
+        if adapter_name in self.reft_A.keys():
+            nn.init.kaiming_uniform_(self.reft_A[adapter_name].weight, a=math.sqrt(5))
+
+        # Move new weights to device
+        weight = getattr(self.get_base_layer(), "weight", None)
+        if weight is not None:
+            # the layer is already completely initialized, this is an update
+            if weight.dtype.is_floating_point or weight.dtype.is_complex:
+                self.to(weight.device, dtype=weight.dtype)
+            else:
+                self.to(weight.device)
+        self.set_adapter(self.active_adapters)
+
+    def get_delta_weight(self, adapter_name: str) -> torch.Tensor:
+        raise NotImplementedError
+
+    def forward(self, x: torch.Tensor, *args, **kwargs) -> torch.Tensor:
+        previous_dtype = x.dtype
+
+        if self.disable_adapters:
+            if self.merged:
+                self.unmerge()
+            output = self.base_layer(x, *args, **kwargs)
+        elif self.merged:
+            output = self.base_layer(x, *args, **kwargs)
+        else:
+            result = self.base_layer(x, *args, **kwargs)
+
+            # Execute all the adapters
+            for active_adapter in self.active_adapters:
+                if active_adapter not in self._available_adapters:
+                    continue
+
+                rotate_layer = self.reft_R[active_adapter]
+                learned_source = self.reft_A[active_adapter]
+                dropout = self.dropout[active_adapter]
+
+                result = result.to(torch.float32)
+                if self.first_n[active_adapter] == 0 and self.last_n[active_adapter] == 0:
+                    rotated_base = rotate_layer(result)
+                    offset = torch.matmul((learned_source(result) - rotated_base), rotate_layer.weight)
```

Comment:
Instead of calling `torch.matmul`, can we not call `rotate_layer(learned_source(result) - rotated_base)`? Generally, it's better to use `forward` than to use the parameters directly.